compiler/x/codegen/SIMDTreeEvaluator.cpp

```diff
@@ -0,0 +1,237 @@
+   cg->decReferenceCount(child);
+   return globalReg;
+   }
+
+TR::Register* OMR::X86::TreeEvaluator::SIMDloadEvaluator(TR::Node* node, TR::CodeGenerator* cg)
+   {
+   TR::MemoryReference* tempMR = generateX86MemoryReference(node, cg);
+   tempMR = ConvertToPatchableMemoryReference(tempMR, node, cg);
+   TR::Register* resultReg = cg->allocateRegister(TR_VRF);
+
+   TR_X86OpCodes opCode = BADIA32Op;
+   switch (node->getSize())
+      {
+      case 16:
+         opCode = MOVDQURegMem;
+         break;
+      default:
+         break;
+      }
+   TR_ASSERT(opCode != BADIA32Op, "SIMD Load Evaluator: unsupported fill size");
+
+   TR::Instruction* instr = generateRegMemInstruction(opCode, node, resultReg, tempMR, cg);
+   if (node->getOpCode().isIndirect())
+      cg->setImplicitExceptionPoint(instr);
+   node->setRegister(resultReg);
+   tempMR->decNodeReferenceCounts(cg);
+   return resultReg;
+   }
+
+TR::Register* OMR::X86::TreeEvaluator::SIMDstoreEvaluator(TR::Node* node, TR::CodeGenerator* cg)
+   {
+   TR::Node* valueNode = node->getChild(node->getOpCode().isIndirect() ? 1 : 0);
+   TR::MemoryReference* tempMR = generateX86MemoryReference(node, cg);
+   tempMR = ConvertToPatchableMemoryReference(tempMR, node, cg);
+   TR::Register* valueReg = cg->evaluate(valueNode);
+
+   TR_X86OpCodes opCode = BADIA32Op;
+   switch (node->getSize())
+      {
+      case 16:
+         opCode = MOVDQUMemReg;
+         break;
+      default:
+         break;
+      }
+   TR_ASSERT(opCode != BADIA32Op, "SIMD Store Evaluator: unsupported fill size");
+
+   TR::Instruction* instr = generateMemRegInstruction(opCode, node, tempMR, valueReg, cg);
+
+   cg->decReferenceCount(valueNode);
+   tempMR->decNodeReferenceCounts(cg);
+   if (node->getOpCode().isIndirect())
+      cg->setImplicitExceptionPoint(instr);
+   return NULL;
+   }
+
+TR::Register* OMR::X86::TreeEvaluator::SIMDsplatsEvaluator(TR::Node* node, TR::CodeGenerator* cg)
+   {
+   TR::Node* childNode = node->getChild(0);
+   TR::Register* childReg = cg->evaluate(childNode);
+
+   uint8_t shufconst = 0;
+   switch (node->getDataType())
+      {
+      case TR::VectorInt32:
+      case TR::VectorFloat:
+         shufconst = 0;
+         break;
+      case TR::VectorInt64:
+      case TR::VectorDouble:
+         shufconst = 0x44;
+         break;
+      default:
+         TR_ASSERT(false, "SIMD Splats Evaluator: unsupported data type");
+         break;
+      }
+
+   TR::Register* resultReg = cg->allocateRegister(TR_VRF);
+   generateRegRegImmInstruction(PSHUFDRegRegImm1, node, resultReg, childReg, shufconst, cg);
+
+   node->setRegister(resultReg);
+   cg->decReferenceCount(childNode);
+   return resultReg;
+   }
+
+enum BinaryArithmetic_t : uint32_t
+   {
+   BinaryArithmeticInvalid,
+   BinaryArithmeticAdd,
+   BinaryArithmeticSub,
+   BinaryArithmeticMul,
+   BinaryArithmeticDiv,
+   BinaryArithmeticNum
+   };
+static const TR_X86OpCodes BinaryArithmeticOpCodes[TR::NumOMRTypes][BinaryArithmeticNum] =
```

Comment:
Fusing the code path also provides future benefits if we want to utilize three-operand AVX instructions. Instead duplicating almost identical code, this evaluator can handle almost, if not all, logic relating to three-operand AVX instructions.